RCTImagePickerManager crash on image from CameraRoll

doochik opened the pull request:
Image from CameraRoll haven't UIImagePickerControllerReferenceURL. So we need to save it to PhotoAlbums first.

To save image I've used the same method from RCTCameraRollManager.

javache commented on Libraries/CameraRoll/RCTImagePickerManager.m:
```diff
@@ -101,9 +104,29 @@ - (dispatch_queue_t)methodQueue
+    UIImage *originalImage = (UIImage *) [info objectForKey: UIImagePickerControllerOriginalImage];
+    NSDictionary *imageMetadata = (NSDictionary *) [info objectForKey: UIImagePickerControllerMediaMetadata];
+
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Isn't this method already called on the main thread?

javache commented on Libraries/CameraRoll/RCTImagePickerManager.m:
```diff
@@ -101,9 +104,29 @@ - (dispatch_queue_t)methodQueue
+    dispatch_async(dispatch_get_main_queue(), ^{
+      [_bridge.assetsLibrary writeImageToSavedPhotosAlbum:originalImage.CGImage metadata:imageMetadata completionBlock:^(NSURL *assetURL, NSError *saveError) {
+        if (saveError) {
+          [self _dismissPicker:picker args:nil];
```
Can we pass the error to JS somehow?

javache commented:
I'm not sure if this is a good default behaviour. In internal projects, we've used RCTImageStoreManager to (temporarily) store images taken this way.

doochik commented:
Yes, this is good point. I've rewrite my PR to use temporary storage

doochik commented:
I've updated my PR. Images from CameraRoll now save to the temporary storage.

And about errors. I've fixed this bug #4411. So now i can pass args something like this. Do you like it?
```
- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary<NSString *, id> *)info
{
  
  NSString *imageUri = [info[UIImagePickerControllerReferenceURL] absoluteString];
  
  // Image from PhotoLibrary
  if (imageUri) {
    [self _dismissPicker:picker args:@[imageUri] error:nil];
    
  } else {
    // Image from CameraRoll hasn't uri.
    // We need to save it to the store first.
    UIImage *originalImage = (UIImage *) [info objectForKey: UIImagePickerControllerOriginalImage];
    [_bridge.imageStoreManager storeImage:originalImage withBlock:^(NSString *tempImageTag) {
      if (!tempImageTag) {
        NSString *errorMessage = @"Error storing image in RCTImageStoreManager";
        RCTLogWarn(@"%@", errorMessage);
        [self _dismissPicker:picker args:nil error:RCTErrorWithMessage(errorMessage)];
        return;
      }
      [self _dismissPicker:picker args:@[tempImageTag] error:nil];
    }];
  }
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
  [self _dismissPicker:picker args:nil error:RCTErrorWithMessage(@"Cancel")];
}
```

nicklockwood commented on Libraries/CameraRoll/RCTImagePickerManager.m:
```diff
@@ -101,9 +104,23 @@ - (dispatch_queue_t)methodQueue
+  } else {
+    // Image from CameraRoll hasn't uri.
+    // We need to save it to the store first.
+    UIImage *originalImage = (UIImage *) [info objectForKey: UIImagePickerControllerOriginalImage];
```
nit: can you use [...] syntax for accessing keys? Also, no need to cast here:
    
     UIImage *originalImage = info[UIImagePickerControllerOriginalImage];

nicklockwood commented:
One minor style issue, but otherwise this is good to go.

javache commented on Libraries/CameraRoll/RCTImagePickerManager.m:
```diff
@@ -101,9 +104,23 @@ - (dispatch_queue_t)methodQueue
+    // Image from CameraRoll hasn't uri.
+    // We need to save it to the store first.
+    UIImage *originalImage = (UIImage *) [info objectForKey: UIImagePickerControllerOriginalImage];
+    [_bridge.imageStoreManager storeImage:originalImage withBlock:^(NSString *tempImageTag) {
```
Can you add a warning here that this will cause a memory-leak as-is? You need to remove the image from the imageStoreManager once you're done using. We're looking into providing automatic solutions for this.

nicklockwood replied:
@javache do you mean a console warning? Or a warning in the documentation?

javache replied:
Just documentation. A warning in the console that you can't fix doesn't sound very useful :)

javache commented:
The error API looks good, but I don't think it's necessary in this case since imageStoreManager doesn't explicitly generate errors. (Any error would/should be a fatal runtime error).

doochik commented:
I've fixed your comments, squash commits and rebase to master

nicklockwood commented:
@facebook-github-bot shipit
